File: services/client/src/protocol/protocol.go
package protocol

import (
	"encoding/binary"
	"io"

	"github.com/7574-sistemas-distribuidos/tp-nivelador/src/safe_socket"
)

const LENGTH_HEADER_SIZE = 4

const FIN_MESSAGE = "FIN"

const ACK_MESSAGE = "OK"

func encodeLength(length int) []byte {
	header := make([]byte, LENGTH_HEADER_SIZE)
	binary.BigEndian.PutUint32(header, uint32(length))
	return header
}

func decodeLength(header []byte) int {
	return int(binary.BigEndian.Uint32(header))
}

func SendMessage(conn io.Writer, payload string) error {
	payloadBytes := []byte(payload)
	header := encodeLength(len(payloadBytes))
	message := append(header, payloadBytes...)
	return safe_socket.SendAll(conn, message)
}

func RecvMessage(conn io.Reader) (string, error) {
	header, err := safe_socket.RecvAll(conn, LENGTH_HEADER_SIZE)
	if err != nil {
		return "", err
	}
	payloadLength := decodeLength(header)

	payloadBytes, err := safe_socket.RecvAll(conn, payloadLength)
	if err != nil {
		return "", err
	}
	return string(payloadBytes), nil
}

File: services/server/src/protocol/protocol.py
import socket
import safe_socket

_LENGTH_HEADER_SIZE = 4

FIN_MESSAGE = "FIN"
ACK_MESSAGE = "OK"


def _read_length_header(sock: socket.socket) -> int:
    header_bytes = safe_socket.recv_all(sock, _LENGTH_HEADER_SIZE)
    return int.from_bytes(header_bytes, byteorder="big")


def recv_message(sock: socket.socket) -> str:
    payload_length = _read_length_header(sock)
    payload_bytes = safe_socket.recv_all(sock, payload_length)
    return payload_bytes.decode("utf-8")


def send_message(sock: socket.socket, payload: str) -> None:
    payload_bytes = payload.encode("utf-8")
    header_bytes = len(payload_bytes).to_bytes(_LENGTH_HEADER_SIZE, byteorder="big")
    safe_socket.send_all(sock, header_bytes + payload_bytes)
